src/algorithms/path_contexts.py

```diff
@@ -0,0 +1,331 @@
+                if not is_noop_line(u) and not is_noop_line(v):
+                    ancestor = lca(u, v)
+                    d = distance(u, v, ancestor)
+                    if d <= max_length:
+                        u, path, v = get_path(u, v, ancestor, leaf_token=leaf_token)
+                        # convert nodes to its desired representation
+                        paths.append((u, [token_extractor(p) for p in path], v))
+
+    return paths
+
+
+# aux function to print paths
+def print_paths(paths, symbols_repr={"UP": "↑", "DOWN": "↓"}):
+    for u, path, v in paths:
+        print(u, end="")
+        for element in path:
+            if type(element) == str and element in symbols_repr.keys():
+                print("%s " % symbols_repr[element], end="")
+            else:
+                print("[%s] " % element, end="")
+
+        print(v)
+
+
+def get_unique_tokens_and_paths(fp):
+    _, paths = fp
+
+    uniq_tokens = set()
+    uniq_paths = set()
+    for u, path, v in paths:
+        uniq_tokens.add(u)
+        uniq_tokens.add(v)
+        uniq_paths.add(path)
+
+    return uniq_tokens, uniq_paths
+
+
+def write_to_disk(output, file_name, paths):
+    # if file_name has relative paths remove them
+    file_name = file_name.replace("../", "")
+    file_out = os.path.join(output, file_name)
+
+    os.makedirs(os.path.dirname(file_out), exist_ok=True)
+
+    print("Writing id paths to %s" % file_out)
+
+    with open(file_out, 'w') as f:
+        for p in paths:
+            f.write(" ".join([str(e) for e in p]) + "\n")
+    f.close()
+
+
+def extract_paths(data, extension, max_length, max_width):
+    """
+    Computes the code paths of each file and the lists of all tokens seen
+    :param data: folder where the AST files are (for now bblfsh binary UASTs)
+    :param extension: extension to filter by; default uast.bin as saved by 'repos_to_uast.py'
+    :param max_length: max length of the code paths
+    :param max_width: max number of leaves between origin and end of a given path
+    :return: id_paths: tuples with file names and its list of path contexts , token2id: all
+    tokens and their id
+    """
+    files = recursive_glob(data, '*%s' % extension)
+
+    file_paths = [None] * len(files)
+    for i, file in enumerate(files):
+        uast = Node.FromString(open(file, 'rb').read())
+
+        p = get_paths(uast, max_length=max_length, max_width=max_width)
+        file_paths[i] = (file, p)
+
+    for file, paths in file_paths:
+        print("\nPaths for file: %s" % file)
+        print_paths(paths)
+
+    # get unique tokens per file
+    tokens_set = [None] * len(files)
+    paths_set = [None] * len(files)
+    for i, fp in enumerate(file_paths):
+        tokens_set[i], paths_set[i] = get_unique_tokens_and_paths(fp)
+
+    # reduce to get global list of unique tokens
+    tokens_set = sorted(set().union(*tokens_set))
+    paths_set = sorted(set().union(*paths_set))
```

Comment:
Continue last comment:

```python

tokens_set = []
paths_set = []
for filename in file_codepaths:
    tokens, paths = get_unique_tokens_and_paths(filename, file_codepaths[filename])
    tokens_set.extend(tokens)
    paths_set.extend(paths)
tokens_set = sorted(set(tokens_set))
paths_set = sorted(set(paths_set))
```

Reply:
why would you do it this way? isn't it just plain longer or is it more efficient/pythonic?